### ISSUE TITLE
t/02-fs-driver.t fails on some systems

### ISSUE THREAD

**Issue by eserte** [doriantaylor/p5-store-digest#1]
On CentOS6 the t/02-fs-driver.t test fails:
```
# Store::Digest::Object
#   Digests:
#     ni:///md5;Xy_GGEhMUqr9J9PLmt43kg
#     ni:///sha-1;8exH4-BWYcmJ2EI-uURgPb3Ck8k
#     ni:///sha-256;o7KyKbiY8_JXzNLEv-xbxWm3OIC5qH9HyhBAWIquuYo
#     ni:///sha-384;lhH3BR8xV3acsDNiuwk1FNeJ2j-m_0agCKjLyo_3ANmGw-eJ66P7H6ZpCgfFJYDB
#     ni:///sha-512;amZx4hPjMQDqGjcb0VUxFpS98fauk0KHP4kR8WlPK1odnjTWVa2UYW72K5sUw-Ml_V_030BztFbIN8uxSKNiYw
#   Size (Bytes): 1489
#   Added to Store: 2018-03-21T13:05:35
#   Last Modified: 2017-09-18T03:34:39
#   Properties Modified: 2018-03-21T13:05:35
#   Content Type: text/plain
#   Validation:
#     content-type    : unverified
#     charset         : unverified
#     content-encoding: unverified
#     syntax          : unverified
Validation failed for type named Num declared in package Specio::Library::Builtins (/opt/perl-5.27.10/lib/site_perl/5.27.10/Specio/Library/Builtins.pm) at line 139 in sub named (eval) with value undef

Trace begun at Specio::Exception->new line 57
Specio::Exception::throw('Specio::Exception', 'message', 'Validation failed for type named Num declared in package Specio::Library::Builtins (/opt/perl-5.27.10/lib/site_perl/5.27.10/Specio/Library/Builtins.pm) at line 139 in sub named (eval) with value undef', 'type', 'Specio::Constraint::Simple=HASH(0x2d9ab68)', 'value', undef) called at (eval 204) line 88
DateTime::_check_from_epoch_params('epoch', undef) called at /opt/perl-5.27.10/lib/site_perl/5.27.10/x86_64-linux/DateTime.pm line 481
DateTime::from_epoch('DateTime', 'epoch', undef) called at /home/cpansand/.cpan/build/2018032112/Store-Digest-0.03-0/blib/lib/Store/Digest/Driver/FileSystem.pm line 1068
Store::Digest::Driver::FileSystem::stats('Store::Digest::Driver::FileSystem=HASH(0x20e3e00)') called at t/02-fs-driver.t line 26
# Looks like your test exited with 255 just after 1.
t/02-fs-driver.t .. 
Dubious, test returned 255 (wstat 65280, 0xff00)
All 1 subtests passed 
```

The test suite also does not pass on FreeBSD systems. See http://matrix.cpantesters.org/?dist=Store-Digest+0.03

**Comment by doriantaylor** [comment]
what is this specio business?

**Comment by doriantaylor** [comment]
ah never mind.

**Comment by doriantaylor** [comment]
OK so this error would come from trying to retrieve ctime/mtime from the control database and for some reason they aren't there, even though they should have been created.

Can you show me what that database looks like by running `db_dump -p t/content/control`?

**Comment by eserte** [comment]
On the freebsd12 smoker it looks like this:
```
$ db_dump-5 -p t/content/control
VERSION=3
format=print
type=hash
db_pagesize=16384
HEADER=END
 algorithms
 md5,sha-1,sha-256,sha-384,sha-512
 ctime
 1521945004
 mtime
 1521945004
 objects
 1
 primary
 sha-256
 bytes
 1515
 deleted
 0
DATA=END
```

**Comment by doriantaylor** [comment]
Thanks for this. I pushed 0.04 with a `diag()` in the beginning of the test and all the ones that fail are DB 4.x while all the ones that pass are ≥ 5.

The `control` database was a `BerkeleyDB::Hash` while all the other metadata is in `BerkeleyDB::Btree`. The only thing I can come up with is either 4.x needs some extra push to commit transactions properly, or the values are getting unset. I just changed that to `Btree` in 0.05 so we'll see if that does anything.

Unfortunately Berkeley DB's documentation is cryptic at best; I got most of this to work in the first place by looking at the BerkeleyDB test suite.
